ts/receiver/mediumGroups.ts

```diff
@@ -82,107 +172,244 @@ async function handleNewGroup(
-          const keyIndex = senderKey.keyIndex as number;
-          await SenderKeyAPI.saveSenderKeys(
-            groupId,
-            PubKey.cast(pubKey),
-            chainKey,
-            keyIndex
-          );
-        } else {
-          log.error('Received invalid sender key');
-        }
-      })
+  // ***** Creating a new group *****
+  log.info('Received a new medium group:', groupId);
+
+  // TODO: Check that we are even a part of this group
+
+  convo.set('name', name);
+  convo.set('members', members);
+  convo.set('is_medium_group', true);
+  convo.set('active_at', Date.now());
+
+  // We only set group admins on group creation
+  convo.set('groupAdmins', admins);
+
+  convo.commit();
+
+  const secretKeyHex = toHex(groupPrivateKey);
+
+  await window.Signal.Data.createOrUpdateIdentityKey({
+    id: groupId,
+    secretKey: secretKeyHex,
+  });
+
+  // Sanity check: we expect to find our own sender key in the list
+  await checkOwnSenderKeyPresent(senderKeys);
+
+  await saveIncomingRatchetKeys(groupId, senderKeys);
+
+  window.SwarmPolling.addGroupId(PubKey.cast(groupId));
+
+  await removeFromCache(envelope);
+}
+
+function sanityCheckMediumGroupUpdate(
+  primary: PubKey,
+  diff: SenderKeyAPI.MemberChanges,
+  groupUpdate: SignalService.MediumGroupUpdate
+) {
+  const joining = diff.joiningMembers || [];
+  const leaving = diff.leavingMembers || [];
+
+  // 1. When there are no member changes, we don't expect any sender keys
+  if (!joining.length && !leaving.length) {
+    if (groupUpdate.senderKeys.length) {
+      window.log.error('Unexpected sender keys in group update');
+    }
+  }
+
+  // 2. With leaving members, we expect keys for all members
+  // (ignoring multidevice for now)
+  if (leaving.length) {
+    const stillMember = leaving.indexOf(primary.key) === -1;
+
+    if (!stillMember) {
+      // Should not receive any sender keys
+      if (groupUpdate.senderKeys.length) {
+        window.log.error('Unexpected sender keys for a leaving member');
+      }
+    } else if (groupUpdate.senderKeys.length < groupUpdate.members.length) {
+      window.log.error('Too few sender keys in group update');
+    }
+  }
+}
+
+async function handleMediumGroupChange(
+  envelope: EnvelopePlus,
+  groupUpdate: SignalService.MediumGroupUpdate
+) {
+  const senderIdentity = envelope.source;
+
+  const {
+    name,
+    groupPublicKey,
+    members: membersBinary,
+    admins: adminsBinary,
+    senderKeys,
+  } = groupUpdate;
+  const { log } = window;
+
+  const groupId = toHex(groupPublicKey);
+
+  const maybeConvo = await window.ConversationController.get(groupId);
+
+  if (!maybeConvo) {
+    log.warn(
+      'Ignoring a medium group update message (INFO) for a non-existing group'
     );
+    await removeFromCache(envelope);
+    // TODO: In practice we probably need to be able to request the group's
+    // the NEW message if we somehow missed the initial group invitation
+    return;
+  }
```

Comment:
maybe we could use requestGroupInfo there?

Reply:
not sure this is handled correctly though

Reply:
Yeah, I don't think it will work correctly, so I left it as a TODO for now.